route/build/src/accum_runoff.f90

```diff
@@ -1,147 +1,64 @@
- character(len=strLen)                          :: cmessage        ! error message from subroutines
-
- ierr=0; message='accum_runoff/'
-
- if (size(NETOPO_in)/=size(RCHFLX_out(iens,:))) then
-  ierr=20; message=trim(message)//'sizes of NETOPO and RCHFLX mismatch'; return
- endif
-
- nSeg = size(RCHFLX_out(iens,:))
-
- allocate(doRoute(nSeg), stat=ierr)
- if(ierr/=0)then; message=trim(message)//'unable to allocate space for [doRoute]'; return; endif
-
- ! if a subset of reaches is processed
- if (present(ixSubRch))then
-   doRoute(:)=.false.
-   doRoute(ixSubRch) = .true. ! only subset of reaches are on
- ! if all the reaches are processed
- else
-   doRoute(:)=.true. ! every reach is on
- endif
-
- nDom = size(river_basin)
+type, extends(base_route_rch) :: accum_route_rch
+ CONTAINS
+   procedure, pass :: route => accum_inst_runoff
+end type accum_route_rch
 
- call t_startf('route/accum-runoff')
-
- do ix = 1,nDom
-   ! 1. Route tributary reaches (parallel)
-   ! compute the sum of all upstream runoff at each point in the river network
-   nTrib=size(river_basin(ix)%branch)
-
-!$OMP PARALLEL DO schedule(dynamic,1)         &
-!$OMP          private(jSeg, iSeg)            & ! private for a given thread
-!$OMP          private(ierr, cmessage)        & ! private for a given thread
-!$OMP          shared(river_basin)            & ! data structure shared
-!$OMP          shared(doRoute)                & ! data array shared
-!$OMP          shared(NETOPO_in)              & ! data structure shared
-!$OMP          shared(RCHFLX_out)             & ! data structure shared
-!$OMP          shared(ix, iEns, ixDesire)     & ! indices shared
-!$OMP          firstprivate(nTrib)
-   do iTrib = 1,nTrib
-     do iSeg=1,river_basin(ix)%branch(iTrib)%nRch
-       jSeg = river_basin(ix)%branch(iTrib)%segIndex(iSeg)
-
-       if (.not. doRoute(jSeg)) cycle
-
-       call accum_qupstream(iens, jSeg, ixDesire, NETOPO_in, RCHFLX_out, ierr, cmessage)
-       if(ierr/=0) call handle_err(ierr, trim(message)//trim(cmessage))
-
-     end do
-   end do
-!$OMP END PARALLEL DO
-
- end do
-
- call t_stopf('route/accum-runoff')
-
- END SUBROUTINE accum_runoff
+CONTAINS
 
  ! *********************************************************************
  ! subroutine: perform accumulate immediate upstream flow
  ! *********************************************************************
- SUBROUTINE accum_qupstream(iEns,       &    ! input: index of runoff ensemble to be processed
-                            segIndex,   &    ! input: index of reach to be processed
-                            ixDesire,   &    ! input: reachID to be checked by on-screen pringing
-                            NETOPO_in,  &    ! input: reach topology data structure
-                            RCHFLX_out, &    ! inout: reach flux data structure
-                            ierr, message)   ! output: error control
+ SUBROUTINE accum_inst_runoff(this,         &
```

Comment:
Here and below I'd add something to say that the "this" is the object this subroutine operates on. Because this is a standard pattern it's not too important, but could be helpful for someone not familiar with object oriented programming. I tend to like more comments in general...

This could be done other places as well, and should just be consistent if it's done...

Reply:
I can add something like this "accum_runoff_rchr" object to bound this procedure